README.md
node-statsd-client
==================

Node.js client for [statsd](https://github.com/etsy/statsd).

Quick tour
----------

```javascript
var SDC = require('statsd-client'),
	sdc = new SDC({host: 'statsd.example.com'});

var timer = new Date();
sdc.increment('some.counter'); // Increment by one.
sdc.gauge('some.gauge', 10); // Set gauge to 10
sdc.timing('some.timer', timer); // Calculates time diff
sdc.histogram('some.histogram', 10, {foo: 'bar'}) // Histogram with tags

sdc.close(); // Optional - stop NOW
```

API
---

### Initialization

```javascript
var SDC = require('statsd-client'),
	sdc = new SDC({host: 'statsd.example.com', port: 8124});
```

Global options:
 * `prefix`: Prefix all stats with this value (default `""`).
 * `tcp`: User specifically wants to use tcp (default `false`).
 * `socketTimeout`: Dual-use timer. Will flush metrics every interval. For UDP,
   it auto-closes the socket after this long without activity (default 1000 ms;
   0 disables this). For TCP, it auto-closes the socket after `socketTimeoutsToClose` number of timeouts have elapsed without activity.
 * `tags`: Object of string key/value pairs which will be appended on to all StatsD payloads (excluding raw payloads) (default `{}`)

UDP options:
 * `host`: Where to send the stats (default `localhost`).
 * `port`: Port to contact the statsd-daemon on (default `8125`).
 * `ipv6`: Use IPv6 instead of IPv4 (default `false`).

TCP options:
 * `host`: Where to send the stats (default `localhost`).
 * `port`: Port to contact the statsd-daemon on (default `8125`).
 * `socketTimeoutsToClose`: Number of timeouts in which the socket auto-closes if it has been inactive. (default `10`; `1` to auto-close after a single timeout).

HTTP options:
 * `host`: The URL to send metrics to (default: `http://localhost`).
 * `headers`: Additional headers to send (default `{}`)
 * `method`: What HTTP method to use (default `PUT`)

To debug, set the environment variable `NODE_DEBUG=statsd-client` when running your program.

### Counting stuff

Counters are supported, both as raw `.counter(metric, delta)` and with the
shortcuts `.increment(metric, [delta=1])` and `.decrement(metric, [delta=-1])`:

```javascript
sdc.increment('systemname.subsystem.value'); // Increment by one
sdc.decrement('systemname.subsystem.value', -10); // Decrement by 10
sdc.counter('systemname.subsystem.value', 100); // Increment by 100
```

### Gauges

Sends an arbitrary number to the back-end:

```javascript
sdc.gauge('what.you.gauge', 100);
sdc.gaugeDelta('what.you.gauge', 20);  // Will now count 120
sdc.gaugeDelta('what.you.gauge', -70); // Will now count 50
sdc.gauge('what.you.gauge', 10);       // Will now count 10
```

### Sets

Send unique occurences of events between flushes to the back-end:

```javascript
sdc.set('your.set', 200);
```

### Delays

Keep track of how fast (or slow) your stuff is:

```javascript
var start = new Date();
setTimeout(function () {
	sdc.timing('random.timeout', start);
}, 100 * Math.random());
```

If it is given a `Date`, it will calculate the difference, and anything else
will be passed straight through.

And don't let the name (or nifty interface) fool you - it can measure any kind
of number, where you want to see the distribution (content lengths, list items,
query sizes, ...)

### Histogram

Many implementations (though not the official one from Etsy) support
histograms as an alias/alternative for timers. So aside from the fancy bits
with handling dates, this is much the same as `.timing()`.


### Raw

Passes a raw string to the underlying socket. Useful for dealing with custom
statsd-extensions in a pinch.

```javascript
sdc.raw('foo.bar:123|t|@0.5|#key:value');
```

### Tags

All the methods above support metric level tags as their last argument. Just like global tags, the format for metric tags is an object of string key/value pairs.
Tags at the metric level overwrite global tags with the same key.

```javascript
sdc.gauge('gauge.with.tags', 100, {foo: 'bar'});
```

### Express helper

There's also a helper for measuring stuff in [Express.js](http://expressjs.com)
via middleware:

```javascript
var SDC = require('statsd-client');
var app = express();
	sdc = new SDC({...});

app.use(sdc.helpers.getExpressMiddleware('somePrefix'));
// or
app.get('/',
	sdc.helpers.getExpressMiddleware('otherPrefix'),
	function (req, res, next) { req.pipe(res); });

app.listen(3000);
```

This will count responses by status-code (`prefix.<statuscode>`) and the
overall response-times.

It can also measure per-URL (e.g. PUT to `/:user/:thing` will become
`PUT_user_thing` by setting the `timeByUrl: true` in the `options`-object:

```javascript
app.use(sdc.helpers.getExpressMiddleware('prefix', { timeByUrl: true }));
```

As the names can become rather odd in corner-cases (esp. regexes and non-REST
interfaces), you can specify another value by setting `res.locals.statsdUrlKey`
at a later point.

The `/` page will appear as `root` (e.g. `GET_root`) in metrics while any not found route will appear as `{METHOD}_unknown_express_route`. You can change that name by setting the `notFoundRouteName` in the middleware options.


### Callback helper

There's also a helper for measuring stuff with regards to a callback:

```javascript
var SDC = requrire('statsd-client');
	sdc = new SDC({...});

function doSomethingAsync(arg, callback) {
	callback = sdc.helpers.wrapCallback('somePrefix', callback);
	// ... do something ...
	return callback(null);
}
```

The callback is overwritten with a shimmed version that counts the
number of errors (`prefix.err`) and successes (`prefix.ok`) and
the time of execution of the function (`prefix.time`).
You invoked the shimmed callback exactly the same way as though
there was no shim at all. Yes, you get metrics for your function in
a single line of code.

Note that the start of execution time is marked as soon as you
invoke `sdc.helpers.wrapCallback()`.

You can also provide more options:

```javascript
sdc.helpers.wrapCallback('somePrefix', callback, {
	tags: { foo: 'bar' }
});
```

### Stopping gracefully

By default, the socket is closed if it hasn't been used for a second (see
`socketTimeout` in the init-options), but it can also be force-closed with
`.close()`:

```javascript
var start = new Date();
setTimeout(function () {
	sdc.timing('random.timeout', start); // 2 - implicitly re-creates socket.
	sdc.close(); // 3 - Closes socket after last use.
}, 100 * Math.random());
sdc.close(); // 1 - Closes socket early.
```

The call is idempotent, so you can call it "just to be sure". And if you submit
new metrics later, the socket will automatically be re-created, and a new
timeout-timer started.

### Prefix-magic

The library supports getting "child" clients with extra prefixes, to help with
making sane name-spacing in apps:

```javascript
// Create generic client
var sdc = new StatsDClient({host: 'statsd.example.com', prefix: 'systemname'});
sdc.increment('foo'); // Increments 'systemname.foo'
... do great stuff ...

// Subsystem A
var sdcA = sdc.getChildClient('a');
sdcA.increment('foo'); // Increments 'systemname.a.foo'

// Subsystem B
var sdcB = sdc.getChildClient('b');
sdcB.increment('foo'); // Increments 'systemname.b.foo'
```

Internally, they all use the same socket, so calling `.close()` on any of them
will allow the entire program to stop gracefully.

What's broken
-------------

Check the [GitHub issues](https://github.com/msiebuhr/node-statsd-client/issues).

Other resources
---------------

 * [statsd-tail](https://github.com/msiebuhr/statsd-tail) - A simple program to grab statsd-data on localhost
 * [hot-shots](https://www.npmjs.com/package/hot-shots) - Another popular statsd client for Node.js
 * [statsd](https://github.com/etsy/statsd) - The canonical server

LICENSE
-------

ISC - see
[LICENSE](https://github.com/msiebuhr/node-statsd-client/blob/master/LICENSE).


CHANGELOG.md
## [1.0.0](https://github.com/mixmaxhq/node-statsd-client/compare/v0.4.3...v1.0.0) (2019-12-10)


### ⚠ BREAKING CHANGES

* Drop support for Node <= 6.

### Features

* improve type inferability ([f40a8b4](https://github.com/mixmaxhq/node-statsd-client/commit/f40a8b451e2cfa6ba08a0c1e74537f9e9d4a69eb))


### Continuous Integration

* remove support for node 6 ([4a92283](https://github.com/mixmaxhq/node-statsd-client/commit/4a922830ab07be19d3e9f49bd52388e6c5e0b85a))


test/helpers.js
const StatsDClient = require('../lib/statsd-client'),
  FakeServer = require('./FakeServer'),
  assert = require('chai').assert,
  express = require('express'),
  supertest = require('supertest'),
  sinon = require('sinon');

/*global after before*/

describe('Helpers', function() {
  let c;
  let s;
  let es;
  let baseUrl;

  before(function(done) {
    s = new FakeServer();
    c = new StatsDClient({
      maxBufferSize: 0,
    });

    const app = express();

    app.use(c.helpers.getExpressMiddleware('express', { timeByUrl: true }));

    // Routes defined on the express app itself.
    app.get('/', function(req, res) {
      res.sendStatus(200);
    });

    app.get('/foo', function(req, res) {
      res.sendStatus(200);
    });

    app.post('/foo', function(req, res) {
      res.sendStatus(200);
    });

    app.get('/foo/:param/bar', function(req, res) {
      res.sendStatus(200);
    });

    // Routes defined on the a subrouter or "micro-app".
    const router = express.Router({ mergeParams: true });

    router.get('/foo', function(req, res) {
      res.sendStatus(200);
    });

    router.get('/foo/:subparam', function(req, res) {
      res.sendStatus(200);
    });

    app.use('/subrouter/:param', router);

    es = app.listen(3000, function() {
      baseUrl = 'http://localhost:' + 3000;
      s.start(done);
    });
  });

  after(function(done) {
    es.on('close', function() {
      s.stop();
      done();
    });
    es.close();
  });

  it('.helpers is an object', function() {
    assert.isObject(c.helpers);
  });

  it('.getExpressMiddleware(prefix) → function (err, res, next)', function() {
    const f = c.helpers.getExpressMiddleware('prefix');
    assert.isFunction(f);
    assert.lengthOf(f, 3);
  });

  describe('response times', function() {
    let sandbox;
    beforeEach(function() {
      sandbox = sinon.sandbox.create();
      sandbox.useFakeTimers(new Date().valueOf(), 'Date');
    });

    afterEach(function() {
      sandbox.restore();
    });

    describe('GET', function() {
      it('should count the response code', function(done) {
        supertest(baseUrl)
          .get('/')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_code.200:1|c', done);
          });
      });

      it('should count the response code with the url prefix', function(done) {
        supertest(baseUrl)
          .get('/')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_code.GET_root.200:1|c', done);
          });
      });

      it('/ → "GET_root"', function(done) {
        supertest(baseUrl)
          .get('/')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_time.GET_root:0|ms', done);
          });
      });

      it('/foo → "GET_foo"', function(done) {
        supertest(baseUrl)
          .get('/foo')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_time.GET_foo:0|ms', done);
          });
      });

      it('/foo/:param/bar → "GET_foo_param_bar"', function(done) {
        supertest(baseUrl)
          .get('/foo/mydynamicparameter/bar')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_time.GET_foo_param_bar:0|ms', done);
          });
      });

      describe('sub-router', function() {
        it('/subrouter/:param/foo → "GET_subrouter_param_foo"', function(done) {
          supertest(baseUrl)
            .get('/subrouter/test/foo')
            .expect(200)
            .end(function(err /*, res*/) {
              if (err) return void done(err);
              s.expectMessage('express.response_time.GET_subrouter_param_foo:0|ms', done);
            });
        });

        it('/subrouter/:param/foo/:subparam → "GET_subrouter_param_foo_subparam"', function(done) {
          supertest(baseUrl)
            .get('/subrouter/test_param/foo/test_sub_param')
            .expect(200)
            .end(function(err /*, res*/) {
              if (err) return void done(err);
              s.expectMessage('express.response_time.GET_subrouter_param_foo_subparam:0|ms', done);
            });
        });
      });
    });

    describe('POST', function() {
      it('/foo → "POST_foo"', function(done) {
        supertest(baseUrl)
          .post('/foo')
          .expect(200)
          .end(function(err /*, res*/) {
            if (err) return void done(err);
            s.expectMessage('express.response_time.POST_foo:0|ms', done);
          });
      });
    });
  });

  it('.wrapCallback(prefix, callback, options) → function', function() {
    const callback = function() {};
    const f = c.helpers.wrapCallback('prefix', callback);
    assert.isFunction(f);
  });

  describe('wrapped callback', function() {
    let sandbox;
    beforeEach(function() {
      sandbox = sinon.sandbox.create();
      sandbox.useFakeTimers(new Date().valueOf(), 'Date');
    });

    afterEach(function() {
      sandbox.restore();
    });

    it('invokes original callback', function(done) {
      const e = new Error('test'),
        v = 20;
      const callback = function(arg1, arg2) {
        assert.strictEqual(arg1, e);
        assert.strictEqual(arg2, v);
        return done();
      };
      const f = c.helpers.wrapCallback('prefix', callback);
      f(e, v);
    });

    describe('sends metrics', function() {
      const callback = function() {};
      let f;
      before(function() {
        f = c.helpers.wrapCallback('callback', callback, {
          tags: { foo: 'bar' },
        });
        f();
      });

      it('sends timing', function(done) {
        s.expectMessage('callback.time:0|ms|#foo:bar', done);
      });

      it('counts errors', function(done) {
        f(new Error('test'));
        s.expectMessage('callback.err:1|c|#foo:bar', done);
      });

      it('counts successes', function(done) {
        s.expectMessage('callback.ok:1|c|#foo:bar', done);
      });
    });
  });
});


test/FakeServer.js
/* A small fake UDP-server.
 */
const dgram = require('dgram');

function FakeServer(options) {
  options = options || {};
  this.port = options.port || 8125;

  this._socket = undefined;
  this._packetsReceived = [];
  this._expectedPackets = [];
}

/* Start the server and listen for messages.
 */
FakeServer.prototype.start = function(cb) {
  const that = this;
  this._socket = dgram.createSocket('udp4');

  this._socket.on('message', function(msg /*, rinfo*/) {
    //console.warn("Server got: '" + msg.toString() + "'");
    msg
      .toString()
      .split('\n')
      .forEach(function(part) {
        that._packetsReceived.push(part);
      });
    that.checkMessages();
  });

  this._socket.on('listening', cb);

  this._socket.bind(this.port);
};

/* For closing server down after use.
 */
FakeServer.prototype.stop = function() {
  this._socket.close();
  this._socket = undefined;
};

/* Expect `message` to arrive and call `cb` if/when it does.
 */
FakeServer.prototype.expectMessage = function(message, cb) {
  const that = this;
  this._expectedPackets.push({
    message,
    callback: cb,
  });
  process.nextTick(function() {
    that.checkMessages();
  });
};

/* Check for expected messages.
 */
FakeServer.prototype.checkMessages = function() {
  const that = this;
  this._expectedPackets.forEach(function(details, detailIndex) {
    // Is it in there?
    const i = that._packetsReceived.indexOf(details.message);
    if (i !== -1) {
      // Remove message and the listener from their respective lists
      that._packetsReceived.splice(i, 1);
      that._expectedPackets.splice(detailIndex, 1);
      details.callback();
    }
  });
};

module.exports = FakeServer;


test/HttpSocket.js
/* Test the Http socket
 */

const HttpSocket = require('../lib/HttpSocket'),
  http = require('http'),
  MessageCollector = require('./messageCollector'),
  assert = require('chai').assert;

/*global before after*/

describe('HttpSocket', function() {
  let s, e, messages, lastHeaders;

  before(function(done) {
    e = new HttpSocket({ host: 'http://localhost:8125' });
    lastHeaders = null;
    messages = new MessageCollector();
    s = http.createServer(function(req, res) {
      lastHeaders = req.headers;
      req.setEncoding('ascii');
      let m = '';
      req.on('data', function(data) {
        m += data;
      });
      req.on('end', function(data) {
        if (data) {
          m += data;
        }
        messages.addMessage(m);
        res.end();
      });
    });
    s.listen(8125, undefined, undefined, done);
  });

  after(function() {
    s.close();
  });

  it('Respects host-configuration', function(done) {
    const w = new HttpSocket({ host: 'some_other_host.sbhr.dk' });
    w.send('wrong_message');

    setTimeout(function() {
      assert.lengthOf(messages._packetsReceived, 0);
      done();
    }, 25);
  });

  it('Sends data immediately with maxBufferSize = 0', function(done) {
    const withoutBuffer = new HttpSocket({ maxBufferSize: 0, host: 'http://localhost:8125' }),
      start = Date.now();

    withoutBuffer.send('do_not_buffer');

    messages.expectMessage(
      'do_not_buffer',
      function() {
        assert.closeTo(Date.now() - start, 0, 100);
        withoutBuffer.close();
        done();
      },
      500
    );
  });

  it("Doesn't send data immediately with maxBufferSize > 0", function(done) {
    const withBuffer = new HttpSocket({ socketTimeout: 25, host: 'http://localhost:8125' });
    withBuffer.send('buffer_this');
    const start = Date.now();

    messages.expectMessage('buffer_this', function(err) {
      assert.operator(Date.now() - start, '>=', 25);
      withBuffer.close();
      done(err);
    });
  });

  it('Sends headers', function(done) {
    const headers = { 'X-Test': 'Test' };
    const withHeaders = new HttpSocket({ headers, host: 'http://localhost:8125' });
    withHeaders.send('no heders kthxbai');
    messages.expectMessage('no heders kthxbai', function(err) {
      assert.isNotNull(lastHeaders);
      assert.equal(lastHeaders['x-test'], 'Test');
      withHeaders.close();
      done(err);
    });
  });

  it('Send 500 messages', function(done) {
    this.slow(500);

    // Send messages
    for (let i = 0; i < 500; i += 1) {
      e.send('foobar' + i);
    }
    e.close();

    setTimeout(function() {
      // Received some packets
      assert.closeTo(
        messages._packetsReceived.length,
        500, // Should get 500
        5 // ±5
      );
      messages._packetsReceived = [];
      return done();
    }, 25);
  });
});


lib/TCPSocket.js
const net = require('net');
const debug = require('util').debuglog('stats-client');

function TCPSocket(options) {
  options = options || {};

  this._hostname = options.host || 'localhost';
  this._port = options.port || 8125;
  this._flushBufferTimeout = 'socketTimeout' in options ? options.socketTimeout : 1000;
  this._timeoutsToClose = 'socketTimeoutsToClose' in options ? options.socketTimeoutsToClose : 10; // close socket if not used in 10 flush intervals

  // Check https://github.com/etsy/statsd/#multi-metric-packets for advisable sizes.
  this._maxBufferSize = 'maxBufferSize' in options ? options.maxBufferSize : 1200;

  // Set up re-usable socket
  this._socket = undefined; // Store the socket here
  this._socketUsed = false; // Flag if it has been used
  this._socketLastUsed = 0; // How many intervals of timeout since socket has been used
  this._socketTimer = undefined; // Reference to check-timer
  this._buffer = [];
}

/* Dual-use timer.
 *
 * First checks if there is anything in it's buffer that need to be sent. If it
 * is non-empty, it will be flushed. (And thusly, the socket is in use and we
 * stop checking further right away).
 *
 * If there is nothing in the buffer and the socket hasn't been used in the
 * previous interval, close it.
 */
TCPSocket.prototype._socketTimeout = function _socketTimeout() {
  debug('close()');
  // Flush the buffer, if it contain anything.
  if (this._buffer.length > 0) {
    this._flushBuffer();
    return;
  }

  // Is it already closed? -- then stop here
  if (!this._socket) {
    return;
  }

  // Not used?
  if (this._socketUsed === false) {
    this._socketLastUsed++;
    // if not used in many intervals, close it
    if (this._socketLastUsed >= this._timeoutsToClose) {
      this.close();
      return;
    }
  } else {
    this._socketLastUsed = 0;
  }

  // Reset whether its been used
  this._socketUsed = false;
  // Start timer, if we have a positive timeout
  if (this._flushBufferTimeout > 0 && !this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
  }
};

/*
 * Close the socket, if in use and cancel the interval-check, if running.
 */
TCPSocket.prototype.close = function close() {
  debug('close()');
  if (!this._socket) {
    return;
  }

  if (this._buffer.length > 0) {
    this._flushBuffer();
  }

  // Cancel the running timer
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }

  // Wait a tick or two, so any remaining stats can be sent.
  setTimeout(this.kill.bind(this), 10);
};

/* Kill the socket RIGHT NOW.
 */
TCPSocket.prototype.kill = function kill() {
  debug('kill()');
  if (!this._socket) {
    return;
  }

  // Clear the timer and catch any further errors silently
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }
  this._socket.on('error', function() {});

  this._socket.end();
  this._socket = undefined;
};

TCPSocket.prototype._createSocket = function _createSocket(callback) {
  debug('_createSocket()');
  if (this._socket) {
    return void callback();
  }

  this._socket = net.Socket({
    type: 'tcp4',
  });

  // Listen on 'error'-events, so they don't bubble up to the main
  // application. Try closing the socket for now, forcing it to be re-created
  // later.
  this._socket.once('error', this.kill.bind(this));

  // Call on when the socket is ready.
  this._socket.connect(this._port, this._hostname, function() {
    return void callback();
  });

  // Start timer, if we have a positive timeout
  if (this._flushBufferTimeout > 0 && !this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
  }
};

/* Buffer management
 */
TCPSocket.prototype._enqueue = function _enqueue(data) {
  debug('_enqueue(', data, ')');

  if (!this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
  }
  // Empty buffer if it's too full
  if (
    this._buffer.reduce(function(sum, line) {
      return sum + line.length;
    }, 0) > this._maxBufferSize
  ) {
    this._flushBuffer();
  }

  this._buffer.push(data);
};

TCPSocket.prototype._flushBuffer = function _flushBuffer() {
  debug('_flushBuffer() →', this._buffer);
  const that = this;
  this._send(this._buffer, function() {
    that._buffer = [];
  });
};

/* Send data - public interface.
 */
TCPSocket.prototype.send = function send(data) {
  debug('send(', data, ')');
  if (this._maxBufferSize === 0) {
    this._send([data], function() {});
  } else {
    this._enqueue(data);
  }
};

/*
 * Send data.
 */
TCPSocket.prototype._send = function _send(data, callback) {
  debug('_send(', data, ')');
  // If we don't have a socket, or we have created one but it isn't
  // ready yet, we need to enqueue data to send once the socket is ready.
  const that = this;

  //Must be asynchronous
  this._createSocket(that._writeToSocket.bind(that, data, callback));
};

TCPSocket.prototype._writeToSocket = function _writeToSocket(data, callback) {
  // Create message
  // Trailing \n important because socket.write will sometimes concat multiple 'write' calls.
  const message = new Buffer(data.join('\n') + '\n');

  debug(message.toString());
  if (this._socket) {
    this._socketUsed = true;
    this._socket.write(message);
  }

  callback();
};

module.exports = TCPSocket;


lib/HttpSocket.js
let http = require('http');
const urlParse = require('url').parse;
const debug = require('util').debuglog('statsd-client');

function HttpSocket(options) {
  options = options || {};

  this._requestOptions = urlParse(options.host || 'http://localhost/');
  this._requestOptions.method = 'PUT';
  this._requestOptions.headers = options.headers || {};
  this._socketTimeoutMsec = 'socketTimeout' in options ? options.socketTimeout : 1000;

  // Require the correct HTTP library
  if (this._requestOptions.protocol === 'https:') {
    http = require('https');
  }

  this._maxBufferSize = 'maxBufferSize' in options ? options.maxBufferSize : 10000;

  // Set up re-usable socket
  this._socketTimer = undefined; // Reference to check-timer
  this._buffer = '';
}

/* Checks if there is anything in it's buffer that need to be sent. If it is
 * non-empty, it will be flushed.
 */
HttpSocket.prototype._socketTimeout = function _socketTimeout() {
  debug('_socketTimeout()');
  // Flush the buffer, if it contain anything.
  if (this._buffer.length > 0) {
    this._flushBuffer();
    return;
  }
};

/*
 * Flush all current data and stop any timers running.
 */
HttpSocket.prototype.close = function close() {
  debug('close()');
  if (this._buffer.length > 0) {
    this._flushBuffer();
  }

  // Cancel the running timer
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }

  // Wait a tick or two, so any remaining stats can be sent.
  setTimeout(this.kill.bind(this), 10);
};

/* Kill the socket RIGHT NOW.
 */
HttpSocket.prototype.kill = function kill() {
  debug('kill()');

  // Clear the timer and catch any further errors silently
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }
};

/* Buffer management
 */
HttpSocket.prototype._enqueue = function _enqueue(data) {
  debug('_enqueue(', data, ')');

  if (!this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._socketTimeoutMsec);
  }

  // Empty buffer if it's too full
  if (this._buffer.length + data.length > this._maxBufferSize) {
    this._flushBuffer();
  }

  if (this._buffer.length === 0) {
    this._buffer = data;
  } else {
    this._buffer += '\n' + data;
  }
};

HttpSocket.prototype._flushBuffer = function _flushBuffer() {
  debug('_flushBuffer() →', this._buffer);
  this._send(this._buffer);
  this._buffer = '';
};

/* Send data - public interface.
 */
HttpSocket.prototype.send = function send(data) {
  debug('send(', data, ')');
  if (this._maxBufferSize === 0) {
    this._send(data);
  } else {
    this._enqueue(data);
  }
};

/*
 * Send data.
 */
HttpSocket.prototype._send = function _send(data) {
  debug('_send(', data, ')');
  const req = http.request(this._requestOptions);

  // Catch but ignore errors
  req.once('error', function() {});

  // Send data
  req.end(data);

  debug(data);
};

module.exports = HttpSocket;


lib/helpers/getExpressMiddleware.js
// Removes ":", heading/trailing / and replaces / by _ in a given route name
function sanitize(routeName) {
  return routeName
    .replace(/:/g, '')
    .replace(/^\/|\/$/g, '')
    .replace(/\//g, '_');
}

// Extracts a route name from the request or response
function findRouteName(req, res) {
  // Did we get a hardcoded name, or should we figure one out?
  if (res.locals && res.locals.statsdUrlKey) {
    return res.locals.statsdUrlKey;
  }

  if (req.route && req.route.path) {
    let routeName = req.route.path;

    if (Object.prototype.toString.call(routeName) === '[object RegExp]') {
      routeName = routeName.source;
    }

    if (req.baseUrl) {
      routeName = req.baseUrl + routeName;
    } else if (routeName === '/') {
      routeName = 'root';
    }

    if (req.params) {
      Object.keys(req.params).forEach(function(key) {
        if (req.params[key] === '') return;
        routeName = routeName.replace(req.params[key], ':' + key);
      });
    }

    // Appends the HTTP method
    return req.method + '_' + sanitize(routeName);
  }

  const undef = undefined;
  return undef;
}

/*
 * Return express middleware that measures overall performance.
 *
 * The `prefix` defaults to `''` (but is currently mandatory). The
 * `options`-argument is optional.
 *  * You can set `timeByUrl`, that add a timer per URL template (ex:
 *    `/api/:username/:thingie`). This can be changed run-time by setting
 *    `res.locals.statsdUrlKey`.
 *  * Add a `function(client, startTime, req, res)` in `onResponseEnd` that
 *    will be called at the very end.
 */
function factory(parentClient) {
  return function(prefix, options) {
    options = options || {};

    const client = parentClient.getChildClient(prefix || '');
    const timeByUrl = options.timeByUrl || false;
    const notFoundRouteName = options.notFoundRouteName || 'unknown_express_route';
    const onResponseEnd = options.onResponseEnd || undefined;

    return function(req, res, next) {
      const startTime = new Date();

      // Shadow end request
      const end = res.end;
      res.end = function() {
        res.end = end;
        end.apply(res, arguments);

        let urlPrefix = '';

        // Time by URL?
        if (timeByUrl) {
          urlPrefix += '.';
          urlPrefix += findRouteName(req, res) || notFoundRouteName;
          client.increment('response_code' + urlPrefix + '.' + res.statusCode);
        }

        client.increment('response_code.' + res.statusCode);
        client.timing('response_time' + urlPrefix, startTime);

        if (onResponseEnd) {
          onResponseEnd(client, startTime, req, res);
        }
      };
      next();
    };
  };
}

module.exports = factory;


lib/EphemeralSocket.js
const dgram = require('dgram');
const debug = require('util').debuglog('statsd-client');

function EphemeralSocket(options) {
  options = options || {};

  this._hostname = options.host || 'localhost';
  this._port = parseInt(options.port, 10) || 8125;
  this._sockettype = options.ipv6 ? 'udp6' : 'udp4';
  this._socketTimeoutMsec = 'socketTimeout' in options ? options.socketTimeout : 1000;

  // Check https://github.com/etsy/statsd/blob/master/docs/metric_types.md#multi-metric-packets for advisable sizes.
  this._maxBufferSize = 'maxBufferSize' in options ? options.maxBufferSize : 1200;

  // Set up re-usable socket
  this._socket = undefined; // Store the socket here
  this._socketUsed = false; // Flag if it has been used
  this._socketTimer = undefined; // Reference to check-timer
  this._buffer = '';
}

EphemeralSocket.prototype.log = function log(/*messages*/) {
  //console.log.apply(null, arguments);
};

/* Dual-use timer.
 *
 * First checks if there is anything in it's buffer that need to be sent. If it
 * is non-empty, it will be flushed. (And thusly, the socket is in use and we
 * stop checking further right away).
 *
 * If there is nothing in the buffer and the socket hasn't been used in the
 * previous interval, close it.
 */
EphemeralSocket.prototype._socketTimeout = function _socketTimeout() {
  debug('close()');
  // Flush the buffer, if it contain anything.
  if (this._buffer.length > 0) {
    this._flushBuffer();
    return;
  }

  // Is it already closed? -- then stop here
  if (!this._socket) {
    return;
  }

  // Has been used? -- reset use-flag and wait some more
  if (this._socketUsed) {
    this._socketUsed = false;
    return;
  }

  // Not used? -- close the socket
  this.close();
};

/*
 * Close the socket, if in use and cancel the interval-check, if running.
 */
EphemeralSocket.prototype.close = function close() {
  debug('close()');
  if (!this._socket) {
    return;
  }

  if (this._buffer.length > 0) {
    this._flushBuffer();
  }

  // Cancel the running timer
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }

  // Wait a tick or two, so any remaining stats can be sent.
  setTimeout(this.kill.bind(this), 10);
};

/* Kill the socket RIGHT NOW.
 */
EphemeralSocket.prototype.kill = function kill() {
  debug('kill()');
  if (!this._socket) {
    return;
  }

  // Clear the timer and catch any further errors silently
  if (this._socketTimer) {
    clearInterval(this._socketTimer);
    this._socketTimer = undefined;
  }
  this._socket.on('error', function() {});

  this._socket.close();
  this._socket = undefined;
};

EphemeralSocket.prototype._createSocket = function _createSocket(callback) {
  debug('_createSocket()');
  if (this._socket) {
    return void callback();
  }

  this._socket = dgram.createSocket(this._sockettype);

  // Listen on 'error'-events, so they don't bubble up to the main
  // application. Try closing the socket for now, forcing it to be re-created
  // later.
  this._socket.once('error', this.kill.bind(this));

  // Call on when the socket is ready.
  this._socket.once('listening', function() {
    return void callback();
  });
  this._socket.bind({ port: 0, exclusive: true }, null);

  // Start timer, if we have a positive timeout
  if (this._socketTimeoutMsec > 0 && !this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._socketTimeoutMsec);
  }
};

/* Buffer management
 */
EphemeralSocket.prototype._enqueue = function _enqueue(data) {
  debug('_enqueue(', data, ')');

  if (!this._socketTimer) {
    this._socketTimer = setInterval(this._socketTimeout.bind(this), this._socketTimeoutMsec);
  }
  // Empty buffer if it's too full
  if (this._buffer.length + data.length > this._maxBufferSize) {
    this._flushBuffer();
  }

  if (this._buffer.length === 0) {
    this._buffer = data;
  } else {
    this._buffer += '\n' + data;
  }
};

EphemeralSocket.prototype._flushBuffer = function _flushBuffer() {
  debug('_flushBuffer() →', this._buffer);
  this._send(this._buffer);
  this._buffer = '';
};

/* Send data - public interface.
 */
EphemeralSocket.prototype.send = function send(data) {
  debug('send(', data, ')');
  if (this._maxBufferSize === 0) {
    this._send(data);
  } else {
    this._enqueue(data);
  }
};

/*
 * Send data.
 */
EphemeralSocket.prototype._send = function _send(data) {
  debug('_send(', data, ')');
  // If we don't have a socket, or we have created one but it isn't
  // ready yet, we need to enqueue data to send once the socket is ready.
  const that = this;

  this._createSocket(function() {
    that._socketUsed = true;

    // Create message
    const message = new Buffer(data);

    debug(message.toString());

    that._socket.send(message, 0, message.length, that._port, that._hostname);
  });
};

module.exports = EphemeralSocket;
